<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>CellMembrane Tutorial • CellMembrane</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="CellMembrane Tutorial">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellMembrane</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CellMembrane-Tutorial.html">CellMembrane Tutorial</a>
    </li>
    <li>
      <a href="../articles/Pseudobulking-and-Filtered-Contrasts.html">Pseudobulking-and-Filtered-Contrasts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BimberLabInternal/CellMembrane/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>CellMembrane Tutorial</h1>
                        <h4 data-toc-skip class="author">GW
McElfresh</h4>
            
            <h4 data-toc-skip class="date">2025-07-10</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BimberLabInternal/CellMembrane/blob/master/vignettes/CellMembrane-Tutorial.Rmd" class="external-link"><code>vignettes/CellMembrane-Tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>CellMembrane-Tutorial.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="software-installation">Software Installation<a class="anchor" aria-label="anchor" href="#software-installation"></a>
</h2>
<p>Several packages exist for single cell RNASeq analysis (R: Seurat,
scran. Python: scanpy) and are generally available for installation
through CRAN or PyPI (i.e. <code>install.packages("Seurat")</code> or
<code>pip install scanpy</code>). We primarily create wrappers for these
software for use in fault-tolerant pipelines and yield useful
visualizations to interpret quality control issues and move onto data
interpretation.</p>
<p>To download our R package <code>CellMembrane</code>, it is
recommended that you create a GitHub account and register your GitHub
credentials in your R Studio session. <code>CellMembrane</code> requires
many other packages from GitHub and, without an account, will exhaust
your allotted anonymous API calls to GitHub. To do this, register a
github account and follow these steps:</p>
<div class="section level3">
<h3 id="create-a-github-account">0: Create a GitHub Account<a class="anchor" aria-label="anchor" href="#create-a-github-account"></a>
</h3>
</div>
<div class="section level3">
<h3 id="install-usethis">1: Install <code>usethis</code><a class="anchor" aria-label="anchor" href="#install-usethis"></a>
</h3>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#use pacman for easy installation from multiple package managers.</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span> <span class="op">(</span><span class="st">"pacman"</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/installed.packages.html" class="external-link">installed.packages</a></span><span class="op">(</span><span class="op">)</span><span class="op">[</span>,<span class="st">"Package"</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">'pacman'</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#load or download the 'usethis' package.</span></span>
<span><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html" class="external-link">p_load</a></span><span class="op">(</span><span class="st">"usethis"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="use-usethis-to-generate-a-personal-access-token-pat">2: Use <code>usethis</code> to Generate a Personal Access Token
(PAT)<a class="anchor" aria-label="anchor" href="#use-usethis-to-generate-a-personal-access-token-pat"></a>
</h3>
<p>This will prompt your browser to open to GitHub, log you in, and land
you on the token generation page. Personal Access Tokens are similar to
a username + password combination, so they can be used to uniquely
identify yourself to GitHub. However, just like a user and password
combination, they are sufficient to perform any actions the PAT has
permissions to perform as your user. Be careful to ensure you treat a
personal access token like you would your password.</p>
<p>Copy this PAT after it is generated.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/github-token.html" class="external-link">create_github_token</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="write-your-pat-into-your--renviron-file">3: Write your PAT into your .Renviron file<a class="anchor" aria-label="anchor" href="#write-your-pat-into-your--renviron-file"></a>
</h3>
<p>The .Reviron file is a (usually hidden) file that R will execute upon
launching. The purpose of this is to establish your credentials each
time you launch R.</p>
<p>There are two ways to store the credentials:</p>
<div class="section level4">
<h4 id="recommended-gitcredsgitcreds_set">Recommended: <code>gitcreds::gitcreds_set()</code><a class="anchor" aria-label="anchor" href="#recommended-gitcredsgitcreds_set"></a>
</h4>
<p>Running <code><a href="https://gitcreds.r-lib.org/reference/gitcreds_get.html" class="external-link">gitcreds::gitcreds_set()</a></code> will prompt you to
paste your github credentials into the RStudio console, which it will
store and use for authentication. Restart RStudio for this to take
effect.</p>
</div>
<div class="section level4">
<h4 id="backup-editing-your--renviron-file">Backup: Editing your .Renviron file<a class="anchor" aria-label="anchor" href="#backup-editing-your--renviron-file"></a>
</h4>
<p>I have experienced that <code><a href="https://gitcreds.r-lib.org/reference/gitcreds_get.html" class="external-link">gitcreds::gitcreds_set()</a></code> is
sometimes unable to locate the correct .Reviron file in less-than-simple
file structures (e.g. on Windows systems with multiple Documents folders
due to OneDrive mirroring). For these cases, it’s more efficient just to
set the PAT yourself by running:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">usethis</span><span class="fu">::</span><span class="fu"><a href="https://usethis.r-lib.org/reference/edit.html" class="external-link">edit_r_environ</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This will open the correct .Renviron file, which you need to paste
your github token into as such:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GITHUB_PAT</span><span class="op">=</span><span class="va">pasteTheGitHubTokenYouGeneratedInStep3Here</span></span></code></pre></div>
<p>Please ensure you have no spaces in line, save the
<code>.Renviron</code> file, restart RStudio, and now you’re ready to
download the Bimber Lab software!</p>
</div>
</div>
<div class="section level3">
<h3 id="initial-cellmembrane-download">4: Initial CellMembrane Download<a class="anchor" aria-label="anchor" href="#initial-cellmembrane-download"></a>
</h3>
<p>There are usually some issues that arise during installation, but to
figure out what these are we need to try to download everything.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html" class="external-link">p_load</a></span><span class="op">(</span><span class="va">BiocManager</span>, <span class="va">devtools</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make sure to update your Rprofile to include Bioconductor repos, such as adding this line to ~/.Rprofile:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/eval.html" class="external-link">local</a></span><span class="op">(</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span>repos <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/repositories.html" class="external-link">repositories</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Latest version:</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">'bimberlabinternal/cellmembrane'</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span>, upgrade <span class="op">=</span> <span class="st">'always'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="iterative-installation-of-problematic-packages">5: Iterative Installation of Problematic Packages<a class="anchor" aria-label="anchor" href="#iterative-installation-of-problematic-packages"></a>
</h3>
<p>You’ll probably experience that DropletUtils and preprocessCore
failed to install, which prevents other software from installing (RIRA,
scuttle, and others.) Installing these packages individually from
Bioconductor usually resolves the issue. You’ll probably be asked to
update packages by typing a/s/ or n. Usually it’s safe to update
packages (option a), but you can reach out to members of the Bimber lab
if issues persist.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#install the generally problematic packages</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DropletUtils"</span>, <span class="st">"preprocessCore"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#try reinstalling the software</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span>repo <span class="op">=</span> <span class="st">'bimberlabinternal/cellmembrane'</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span>, upgrade <span class="op">=</span> <span class="st">'always'</span><span class="op">)</span></span></code></pre></div>
<p>You may need to Google installation instructions for the individual
packages that fail to automatically install, but this iterative
installation is generally a one-time process.</p>
</div>
</div>
<div class="section level2">
<h2 id="the-canonical-pipeline">The Canonical Pipeline<a class="anchor" aria-label="anchor" href="#the-canonical-pipeline"></a>
</h2>
<p>Single cell RNASeq procesing generally consists of three steps:</p>
<ul>
<li>Normalization</li>
<li>Dimensional reduction via PCA</li>
<li>Clustering via a community detection algorithm (Louvain clustering
in our case, Leiden in others.)</li>
</ul>
<p>We can use the Seurat wrappers in CellMembrane to perform these
steps. Each step will show plots that vary from quality control to
high-level data summarizations.</p>
<div class="section level3">
<h3 id="normalize-the-data">1: Normalize The Data<a class="anchor" aria-label="anchor" href="#normalize-the-data"></a>
</h3>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#load Seurat to gain access to pbmc_small</span></span>
<span><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html" class="external-link">p_load</a></span><span class="op">(</span><span class="va">Seurat</span>, <span class="va">CellMembrane</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#Normalize the data. </span></span>
<span><span class="co">#We also perform cell cycle scoring here, but pbmc_small does not have these genes</span></span>
<span><span class="co">#so we'll skip it for the tutorial</span></span>
<span><span class="va">pbmc_small</span> <span class="op">&lt;-</span> <span class="fu">CellMembrane</span><span class="fu">::</span><span class="fu"><a href="../reference/NormalizeAndScale.html">NormalizeAndScale</a></span><span class="op">(</span><span class="va">pbmc_small</span>, scoreCellCycle <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The `slot` argument of `SetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">## Warning: The `slot` argument of `GetAssayData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<pre><code><span><span class="co">## [1] "Find variable features:"</span></span></code></pre>
<pre><code><span><span class="co">## Warning: `invoke()` is deprecated as of rlang 0.4.0.</span></span>
<span><span class="co">## Please use `exec()` or `inject()` instead.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span></code></pre>
<pre><code><span><span class="co">## [1] "Total variable features: 230"</span></span></code></pre>
<p><img src="CellMembrane-Tutorial_files/figure-html/NormalizeAndScale-1.png" width="700"></p>
<pre><code><span><span class="co">## [1] "ScaleData will use the top 230 variableFeatures"</span></span>
<span><span class="co">## [1] "Scale data:"</span></span>
<span><span class="co">## [1] "Cell cycle scoring will be skipped"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="run-pca">2: Run PCA<a class="anchor" aria-label="anchor" href="#run-pca"></a>
</h3>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small</span> <span class="op">&lt;-</span> <span class="fu">CellMembrane</span><span class="fu">::</span><span class="fu"><a href="../reference/RunPcaSteps.html">RunPcaSteps</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Total variable genes: 230"</span></span></code></pre>
<pre><code><span><span class="co">## Warning in svd.function(A = t(x = object), nv = npcs, ...): You're computing</span></span>
<span><span class="co">## too large a percentage of total singular values, use a standard svd instead.</span></span></code></pre>
<pre><code><span><span class="co">## Warning: Number of dimensions changing from 19 to 50</span></span></code></pre>
<pre><code><span><span class="co">## PC_ 1 </span></span>
<span><span class="co">## Positive:  IFITM2, S100A11, LGALS1, TYROBP, TYMP, LST1, ANXA2, NFKBIA, CTSS, AIF1 </span></span>
<span><span class="co">##     FCER1G, HLA-DPA1, FCN1, RGS2, IFITM3, EIF3G, BID, YWHAB, GSTP1, HLA-DRB1 </span></span>
<span><span class="co">## Negative:  GNG11, GP9, SDPR, CLU, PF4, HIST1H2AC, SPARC, PPBP, CD9, TUBB1 </span></span>
<span><span class="co">##     CA2, ITGA2B, TMEM40, NGFRAP1, NRGN, TREML1, PGRMC1, PTCRA, MYL9, RUFY1 </span></span>
<span><span class="co">## PC_ 2 </span></span>
<span><span class="co">## Positive:  CST3, LYZ, AIF1, HLA-DRA, LST1, TYMP, HLA-DPB1, IFITM3, HLA-DRB1, CTSS </span></span>
<span><span class="co">##     S100A8, HLA-DRB5, HLA-DPA1, S100A9, SERPINA1, HLA-DQB1, GRN, FCN1, SAT1, FCGRT </span></span>
<span><span class="co">## Negative:  CST7, CTSW, GZMA, GZMM, PRF1, NKG7, GNLY, GZMB, FGFBP2, LCK </span></span>
<span><span class="co">##     GZMH, CD247, CD7, SPON2, LAMP1, AKR1C3, CCL4, KLRD1, C12orf75, RARRES3 </span></span>
<span><span class="co">## PC_ 3 </span></span>
<span><span class="co">## Positive:  LGALS1, FCER1G, TYROBP, TYMP, S100A9, PSAP, S100A8, SAT1, CFD, IFITM3 </span></span>
<span><span class="co">##     SMCO4, CST3, CD68, FCN1, CTSB, S100A11, SERPINA1, LST1, C5AR1, BID </span></span>
<span><span class="co">## Negative:  MS4A1, CD79A, TCL1A, HLA-DQB1, HLA-DQA1, HLA-DMB, FCER2, LY86, PPAPDC1B, CD79B </span></span>
<span><span class="co">##     CYB561A3, KIAA0125, HLA-DQA2, HLA-DRB5, NT5C, HLA-DRB1, CD180, LINC00926, FCRLA, HVCN1 </span></span>
<span><span class="co">## PC_ 4 </span></span>
<span><span class="co">## Positive:  FGFBP2, GZMB, CCL4, AKR1C3, FCER1G, CLIC3, TTC38, GNLY, KLRD1, IL2RB </span></span>
<span><span class="co">##     MS4A1, HLA-DRB1, PRF1, GZMH, IGFBP7, CD79A, TCL1A, ARHGDIA, HLA-DPA1, NKG7 </span></span>
<span><span class="co">## Negative:  CD3D, IL7R, SAFB2, MAL, THYN1, CD2, LDHB, ASNSD1, CD3E, LTB </span></span>
<span><span class="co">##     NOSIP, ACAP1, CCDC104, CCR7, TAGAP, MPHOSPH6, TMEM204, TAF7, DNAJB1, TMEM123 </span></span>
<span><span class="co">## PC_ 5 </span></span>
<span><span class="co">## Positive:  TNFRSF1B, RP11-290F20.3, MS4A7, LILRA3, FCGR3A, RHOC, WARS, CD68, MYO1G, HCK </span></span>
<span><span class="co">##     IFI30, SERPINA1, C5AR1, CD79B, IFITM2, FAM96A, LTB, RGS2, EPC1, COPS6 </span></span>
<span><span class="co">## Negative:  MS4A6A, LGALS2, CLEC10A, FCER1A, IL1B, CD14, S100A8, RGS1, GSTP1, ASGR1 </span></span>
<span><span class="co">##     MPEG1, IL17RA, RBP7, HLA-DMA, S100A9, FUOM, LYZ, ZNF330, NUP214, CTSB</span></span></code></pre>
<pre><code><span><span class="co">## Warning in JackStraw(object = seuratObj, num.replicate = 100, verbose = F):</span></span>
<span><span class="co">## Number of variable genes given 0.01 as the prop.freq is low. Consider including</span></span>
<span><span class="co">## more variable genes and/or increasing prop.freq. Continuing with 3 genes in</span></span>
<span><span class="co">## every random sampling.</span></span></code></pre>
<p><img src="CellMembrane-Tutorial_files/figure-html/RunPcaSteps-1.png" width="700"><img src="CellMembrane-Tutorial_files/figure-html/RunPcaSteps-2.png" width="700"><img src="CellMembrane-Tutorial_files/figure-html/RunPcaSteps-3.png" width="700"></p>
<pre><code><span><span class="co">## Warning: The `slot` argument of `FetchData()` is deprecated as of SeuratObject 5.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use the `layer` argument instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<p><img src="CellMembrane-Tutorial_files/figure-html/RunPcaSteps-4.png" width="700"><img src="CellMembrane-Tutorial_files/figure-html/RunPcaSteps-5.png" width="700"><img src="CellMembrane-Tutorial_files/figure-html/RunPcaSteps-6.png" width="700"><img src="CellMembrane-Tutorial_files/figure-html/RunPcaSteps-7.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="run-clustering-and-umaptsne">3: Run Clustering (and UMAP/tSNE)<a class="anchor" aria-label="anchor" href="#run-clustering-and-umaptsne"></a>
</h3>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc_small</span> <span class="op">&lt;-</span> <span class="fu">CellMembrane</span><span class="fu">::</span><span class="fu"><a href="../reference/FindClustersAndDimRedux.html">FindClustersAndDimRedux</a></span><span class="op">(</span><span class="va">pbmc_small</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "min.y greater than smallest y"</span></span></code></pre>
<p><img src="CellMembrane-Tutorial_files/figure-html/FindClustersAndDimRedux-1.png" width="700"></p>
<pre><code><span><span class="co">## [1] "Inferred elbow: 7"</span></span>
<span><span class="co">## [1] "Selected dimsToUse: 1:7"</span></span>
<span><span class="co">## [1] "Setting RunUMAP arg: umap.method to uwot"</span></span>
<span><span class="co">## [1] "Setting RunUMAP arg: seed.use to 1234"</span></span>
<span><span class="co">## [1] "Setting RunUMAP arg: densmap to FALSE"</span></span></code></pre>
<pre><code><span><span class="co">## Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span></span>
<span><span class="co">## To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span></span>
<span><span class="co">## This message will be shown once per session</span></span></code></pre>
<p><img src="CellMembrane-Tutorial_files/figure-html/FindClustersAndDimRedux-2.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="basic-exploratory-data-investigation">Basic Exploratory Data Investigation<a class="anchor" aria-label="anchor" href="#basic-exploratory-data-investigation"></a>
</h2>
<p>Now, we’re prepared to investigate the data through several
tools.</p>
<div class="section level3">
<h3 id="discrete-data-visualizations">1: Discrete Data Visualizations<a class="anchor" aria-label="anchor" href="#discrete-data-visualizations"></a>
</h3>
<p>For discrete data stored in the <code>pbmc_small@meta.data</code>
data frame (such as clustering identity, subject ID numbers, vaccination
status), we can use Seurat’s <code>DimPlot()</code> to visualize the
data. Clustering identity is computed over several resolutions and
stored in metadata fields named <code>ClusterNames_</code> + a
resolution parameter (increasing in units of 0.2).</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">'groups'</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellMembrane-Tutorial_files/figure-html/DimPlots-1.png" width="700"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">'ClusterNames_0.2'</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellMembrane-Tutorial_files/figure-html/DimPlots-2.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">'ClusterNames_0.8'</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellMembrane-Tutorial_files/figure-html/DimPlots-3.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DimPlot</span><span class="op">(</span><span class="va">pbmc_small</span>, group.by <span class="op">=</span> <span class="st">'ClusterNames_1.2'</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellMembrane-Tutorial_files/figure-html/DimPlots-4.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="continuous-data-visualization">2: Continuous Data Visualization<a class="anchor" aria-label="anchor" href="#continuous-data-visualization"></a>
</h3>
<p>For continuous variables and expression of individual genes, you can
use Seurat’s <code>FeaturePlot()</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">pbmc_small</span>, features <span class="op">=</span> <span class="st">'nCount_RNA'</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellMembrane-Tutorial_files/figure-html/FeaturePlots-1.png" width="700"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">FeaturePlot</span><span class="op">(</span><span class="va">pbmc_small</span>, features <span class="op">=</span> <span class="st">'PPBP'</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellMembrane-Tutorial_files/figure-html/FeaturePlots-2.png" width="700"></p>
<p>We can combine these observations to conclude that Cluster 4 (at
resolution <code>ClusterNames_1.2</code>) are probably GEMs of platelets
or have heavy platelet contamination due to their high expression of
Pro-Platelet basic protein (PPBP).</p>
</div>
</div>
<div class="section level2">
<h2 id="differential-expression">Differential Expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h2>
<p>Take heed that p values derived from cluster-based differential
expression are statistically invalid, as they suffer from double dipping
(please read <a href="https://arxiv.org/abs/2301.07276" class="external-link uri">https://arxiv.org/abs/2301.07276</a> and related works from
the Witten lab). Generally speaking, this means your p values are
approximately zero for large swathes of the transcriptome.</p>
<p>However, for determining primary cluster defining markers, we can
still use these approaches as long as we don’t use thresholds on the p
values to determine differential expression.</p>
<p>To do this, we can use Seurat’s <code>FindMarkers</code>. We need to
specify our groups for differential expression.</p>
<div class="section level3">
<h3 id="targeted-differential-expression">1: Targeted Differential Expression<a class="anchor" aria-label="anchor" href="#targeted-differential-expression"></a>
</h3>
<p>We can perform targeted differential expression (Group 1 = Cluster 4,
Group 2 = Cluster 0 in cluster resolution 1.2) by specifying the
<code>ident.1</code>, <code>ident.2</code>, and the grouping field
<code>group.by</code> arguments.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">FindMarkers</span><span class="op">(</span><span class="va">pbmc_small</span>, ident.1 <span class="op">=</span> <span class="fl">4</span>, ident.2 <span class="op">=</span> <span class="fl">0</span>, group.by <span class="op">=</span> <span class="st">"ClusterNames_1.2"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: `PackageCheck()` was deprecated in SeuratObject 5.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Please use `rlang::check_installed()` instead.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> The deprecated feature was likely used in the <span style="color: #0000BB;">Seurat</span> package.</span></span>
<span><span class="co">##   Please report the issue at <span style="color: #0000BB; font-style: italic;">&lt;https://github.com/satijalab/seurat/issues&gt;</span>.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once every 8 hours.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                  p_val avg_log2FC pct.1 pct.2    p_val_adj</span></span>
<span><span class="co">## GP9       3.055940e-08  12.105216     1 0.000 7.028661e-06</span></span>
<span><span class="co">## PF4       9.206983e-08   9.001797     1 0.042 2.117606e-05</span></span>
<span><span class="co">## SDPR      9.206983e-08   6.794271     1 0.042 2.117606e-05</span></span>
<span><span class="co">## GNG11     9.206983e-08   7.938289     1 0.042 2.117606e-05</span></span>
<span><span class="co">## CLU       9.206983e-08   6.725616     1 0.042 2.117606e-05</span></span>
<span><span class="co">## HIST1H2AC 9.206983e-08   7.013363     1 0.042 2.117606e-05</span></span></code></pre>
<p>Here we can see that GP9 (platelet glycoprotein) has far more
expression in cluster 4 than cluster 0.</p>
</div>
<div class="section level3">
<h3 id="one-versus-all-differential-expression">2: One versus All Differential Expression<a class="anchor" aria-label="anchor" href="#one-versus-all-differential-expression"></a>
</h3>
<p>However, we can also perform a “Cluster 4 versus the rest of the
data” comparison. We can leave <code>ident.2 = NULL</code> to specify
that we want to do this one versus all comparison.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">&lt;-</span> <span class="fu">FindMarkers</span><span class="op">(</span><span class="va">pbmc_small</span>, ident.1 <span class="op">=</span> <span class="fl">4</span>, ident.2 <span class="op">=</span> <span class="cn">NULL</span>, group.by <span class="op">=</span> <span class="st">"ClusterNames_1.2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">markers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##              p_val avg_log2FC pct.1 pct.2    p_val_adj</span></span>
<span><span class="co">## GP9   8.479670e-19  13.670000     1 0.000 1.950324e-16</span></span>
<span><span class="co">## PF4   2.816457e-17  10.566582     1 0.014 6.477852e-15</span></span>
<span><span class="co">## GNG11 2.816457e-17   9.503074     1 0.014 6.477852e-15</span></span>
<span><span class="co">## CLU   2.816457e-17   8.290401     1 0.014 6.477852e-15</span></span>
<span><span class="co">## CD9   3.533280e-17   7.467040     1 0.014 8.126545e-15</span></span>
<span><span class="co">## SDPR  4.928208e-16   8.154001     1 0.028 1.133488e-13</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="marker-sorting-and-interpretation">3: Marker Sorting and Interpretation<a class="anchor" aria-label="anchor" href="#marker-sorting-and-interpretation"></a>
</h3>
<p>By default, the markers are sorted by lowest p value. You can also
sort the data by largest average fold change to find reliable
markers.</p>
<p>For information on the pipe symbol <code>%&gt;%</code> used below,
seek information on ‘tidy data’ from Hadley Wickham and his
<code>tidyverse</code>. (<a href="https://r4ds.hadley.nz/data-tidy.html" class="external-link uri">https://r4ds.hadley.nz/data-tidy.html</a>)</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">pacman</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pacman/man/p_load.html" class="external-link">p_load</a></span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="va">markers</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##               p_val avg_log2FC pct.1 pct.2    p_val_adj</span></span>
<span><span class="co">## GP9    8.479670e-19  13.670000 1.000 0.000 1.950324e-16</span></span>
<span><span class="co">## PTCRA  1.391559e-12  12.954246 0.667 0.000 3.200585e-10</span></span>
<span><span class="co">## TMEM40 1.314337e-14  12.401665 0.778 0.000 3.022976e-12</span></span>
<span><span class="co">## PF4    2.816457e-17  10.566582 1.000 0.014 6.477852e-15</span></span>
<span><span class="co">## GNG11  2.816457e-17   9.503074 1.000 0.014 6.477852e-15</span></span>
<span><span class="co">## CA2    4.017473e-15   8.760577 0.889 0.014 9.240187e-13</span></span></code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">markers</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">avg_log2FC</span><span class="op">)</span> <span class="op">%&gt;%</span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                p_val avg_log2FC pct.1 pct.2 p_val_adj</span></span>
<span><span class="co">## LST1     0.008374342  -10.21478     0 0.493 1.0000000</span></span>
<span><span class="co">## GNLY     0.082814591  -10.44640     0 0.268 1.0000000</span></span>
<span><span class="co">## HLA-DPA1 0.002385780  -10.65441     0 0.592 0.5487293</span></span>
<span><span class="co">## TYROBP   0.001066881  -10.73993     0 0.648 0.2453827</span></span>
<span><span class="co">## HLA-DRB1 0.002887406  -10.77214     0 0.577 0.6641033</span></span>
<span><span class="co">## NKG7     0.008374806  -11.56659     0 0.493 1.0000000</span></span></code></pre>
<p>We can see that GP9, PTCRA, and TMEM40 are likely good positive
markers of our platelet population, while GNLY, LST1, and HLA-DPA1
should be expressed in other clusters Let’s visualize these using other
standard Seurat functions: <code>DotPlot</code> and
<code>VlnPlot</code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DotPlot</span><span class="op">(</span><span class="va">pbmc_small</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GP9"</span>, <span class="st">"PTCRA"</span>, <span class="st">"TMEM40"</span>, <span class="st">"LST1"</span>, <span class="st">"GNLY"</span>, <span class="st">"HLA-DPA1"</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"ClusterNames_1.2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellMembrane-Tutorial_files/figure-html/DotandVlnPlots-1.png" width="700"></p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">VlnPlot</span><span class="op">(</span><span class="va">pbmc_small</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GP9"</span>, <span class="st">"PTCRA"</span>, <span class="st">"TMEM40"</span>, <span class="st">"LST1"</span>, <span class="st">"GNLY"</span>, <span class="st">"HLA-DPA1"</span><span class="op">)</span>, group.by <span class="op">=</span> <span class="st">"ClusterNames_1.2"</span><span class="op">)</span></span></code></pre></div>
<p><img src="CellMembrane-Tutorial_files/figure-html/DotandVlnPlots-2.png" width="700"></p>
<p>Visualizing your data through unsupervised clustering to find sources
of variation in your transcriptomics data is useful and can help you
link changes in gene expression to your experimental variables.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

      </footer>
</div>






  </body>
</html>
