<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Pseudobulking-and-Filtered-Contrasts • CellMembrane</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Pseudobulking-and-Filtered-Contrasts">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CellMembrane</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/CellMembrane-Tutorial.html">CellMembrane Tutorial</a>
    </li>
    <li>
      <a href="../articles/Pseudobulking-and-Filtered-Contrasts.html">Pseudobulking-and-Filtered-Contrasts</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BimberLabInternal/CellMembrane/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Pseudobulking-and-Filtered-Contrasts</h1>
                        <h4 data-toc-skip class="author">GW
McElfresh</h4>
            
            <h4 data-toc-skip class="date">2025-01-29</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BimberLabInternal/CellMembrane/blob/master/vignettes/Pseudobulking-and-Filtered-Contrasts.Rmd" class="external-link"><code>vignettes/Pseudobulking-and-Filtered-Contrasts.Rmd</code></a></small>
      <div class="hidden name"><code>Pseudobulking-and-Filtered-Contrasts.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="what-is-pseudobulking">What is Pseudobulking?<a class="anchor" aria-label="anchor" href="#what-is-pseudobulking"></a>
</h2>
<p>Pseudobulking is a technique used in single cell RNA Sequencing to
simplify the data from many single cells into a (pseudo) bulked sample
by summing all of the gene expression from each cell in the group.
Groupings can be determined several different ways.</p>
<ul>
<li>Technical batching (e.g. a 10X lane or all of the cells from a
single subject).</li>
<li>Biological phenotypes (e.g. T cells vs B cells).</li>
</ul>
<p>Pseudobulking approaches should include groupings closely aligned
with your experimental variables (e.g. pseudobulking on Subject,
Treatment Status, and Timepoint) and generally not be used downstream of
unsupervised clustering for reasons discussed below.</p>
</div>
<div class="section level2">
<h2 id="why-should-we-use-pseudobulking">Why should we use Pseudobulking?<a class="anchor" aria-label="anchor" href="#why-should-we-use-pseudobulking"></a>
</h2>
<p>In scRNASeq analyses, it is very common to use unsupervised
clustering approaches to create clusters of cells for which their gene
expression is grossly different. These cell groupings sometimes
correspond to distinct cellular phenotypes, but other times clusters
capture transient cell states such as division or metabolic shifts. Once
these clusters are obtained, one generally runs a variation of Seurat’s
<code>FindMarkers()</code> to perform a wilcoxon test between two
clusters, or a one-versus-all test to compare a single cluster to the
rest of the data grouped together. This approach works well for
determining the top 10/50/100 genes that define the population, but
falls short when one asks “What are the differentially expressed
genes?”.</p>
<p>Differential gene expression usually considers the “strength” of
expression
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>L</mi><mi>o</mi><msub><mi>g</mi><mn>2</mn></msub><mrow><mo stretchy="true" form="prefix">(</mo><mfrac displaystyle="true"><mrow><mi>E</mi><mi>x</mi><mi>p</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>s</mi><mi>i</mi><mi>o</mi><msub><mi>n</mi><mrow><mi>e</mi><mi>x</mi><mi>p</mi><mi>e</mi><mi>r</mi><mi>i</mi><mi>m</mi><mi>e</mi><mi>n</mi><mi>t</mi><mi>a</mi><mi>l</mi></mrow></msub></mrow><mrow><mi>E</mi><mi>x</mi><mi>p</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>s</mi><mi>i</mi><mi>o</mi><msub><mi>n</mi><mrow><mi>c</mi><mi>o</mi><mi>n</mi><mi>t</mi><mi>r</mi><mi>o</mi><mi>l</mi></mrow></msub></mrow></mfrac><mo stretchy="true" form="postfix">)</mo></mrow></mrow><annotation encoding="application/x-tex">Log_2(\dfrac{Expression_{experimental}}{Expression_{control}})</annotation></semantics></math>
and the p value as computed via a hypothesis test versus a null
distribution. Frequently, you will discover that the p values obtained
via the clustering and <code>FindMarkers()</code> method will be
infinitesimally small. This is due to a phenomenon known as double
dipping, wherein the clustering was determined through differences in
gene expression, and then the genes were tested for “how different is
the gene’s expression between clusters” using a null hypothesis that is
no longer feasible (e.g. using
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mi>x</mi><mi>p</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>s</mi><mi>i</mi><mi>o</mi><msub><mi>n</mi><mrow><mi>c</mi><mi>l</mi><mi>u</mi><mi>s</mi><mi>t</mi><mi>e</mi><msub><mi>r</mi><mn>1</mn></msub></mrow></msub><mo>=</mo><mi>E</mi><mi>x</mi><mi>p</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>s</mi><mi>i</mi><mi>o</mi><msub><mi>n</mi><mrow><mi>c</mi><mi>l</mi><mi>u</mi><mi>s</mi><mi>t</mi><mi>e</mi><msub><mi>r</mi><mn>2</mn></msub></mrow></msub></mrow><annotation encoding="application/x-tex">Expression_{cluster_1} = Expression_{cluster_2}</annotation></semantics></math>
as a null hypothesis when cluster membership maximizes
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>E</mi><mi>x</mi><mi>p</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>s</mi><mi>i</mi><mi>o</mi><msub><mi>n</mi><mrow><mi>c</mi><mi>l</mi><mi>u</mi><mi>s</mi><mi>t</mi><mi>e</mi><msub><mi>r</mi><mn>1</mn></msub></mrow></msub><mo>≠</mo><mi>E</mi><mi>x</mi><mi>p</mi><mi>r</mi><mi>e</mi><mi>s</mi><mi>s</mi><mi>i</mi><mi>o</mi><msub><mi>n</mi><mrow><mi>c</mi><mi>l</mi><mi>u</mi><mi>s</mi><mi>t</mi><mi>e</mi><msub><mi>r</mi><mn>2</mn></msub></mrow></msub></mrow><annotation encoding="application/x-tex">Expression_{cluster_1} \neq Expression_{cluster_2}</annotation></semantics></math>).</p>
<p>Pseudobulking offers a way to statistically determine gene sets that
are generally congruent with one’s expectations of type I and type II
error control.</p>
<p>For more on double dipping and approaches to circumvent it, please
read <a href="https://arxiv.org/abs/2301.07276" class="external-link uri">https://arxiv.org/abs/2301.07276</a> and related works from
the Witten lab.</p>
</div>
<div class="section level2">
<h2 id="how-to-pseudobulk">How to Pseudobulk<a class="anchor" aria-label="anchor" href="#how-to-pseudobulk"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">#Load seurat object. GITHUB_WORKSPACE is set in the GitHub actions build, and if this is not set, try to build relative to this file:</span></span>
<span><span class="va">seuratObj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">suppressMessages</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span></span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/UpdateSeuratObject.html" class="external-link">UpdateSeuratObject</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.getenv.html" class="external-link">Sys.getenv</a></span><span class="op">(</span><span class="st">'GITHUB_WORKSPACE'</span>, unset <span class="op">=</span> <span class="st">'..'</span><span class="op">)</span>, <span class="st">"/tests/testdata/seuratOutput.rds"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#assign some random experimental variables: vaccine groups, experimental time points, and subject Ids. </span></span>
<span><span class="va">seuratObj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="st">"vaccine_cohort"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"control"</span>, <span class="st">"vaccineOne"</span>, <span class="st">"vaccineTwo"</span>, <span class="st">"unvax"</span><span class="op">)</span>, length.out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seuratObj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">seuratObj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="st">"timepoint"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"baseline"</span>, <span class="st">"necropsy"</span>, <span class="st">"day4"</span><span class="op">)</span>, length.out <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">seuratObj</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">seuratObj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">[</span>,<span class="st">"subject"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">base</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span>,<span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">1557</span>, replace <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">seuratObj</span><span class="op">@</span><span class="va">meta.data</span><span class="op">$</span><span class="va">sample_identifier</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Subject-"</span>, <span class="va">seuratObj</span><span class="op">$</span><span class="va">subject</span>, <span class="st">'-'</span>, <span class="va">seuratObj</span><span class="op">$</span><span class="va">vaccine_cohort</span>, <span class="st">"-"</span>, <span class="va">seuratObj</span><span class="op">$</span><span class="va">timepoint</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#set up pseudobulking</span></span>
<span><span class="va">pbulk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="../reference/PseudobulkSeurat.html">PseudobulkSeurat</a></span><span class="op">(</span><span class="va">seuratObj</span>, groupFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vaccine_cohort"</span>, <span class="st">"timepoint"</span>, <span class="st">"sample_identifier"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/start%20pseudobulking-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="inspecting-the-pseudobulked-object-">Inspecting the Pseudobulked Object.<a class="anchor" aria-label="anchor" href="#inspecting-the-pseudobulked-object-"></a>
</h2>
<p>The pseudobulked Seurat object has two layers of interest: the
<code>counts</code> layer and the <code>pct.expression</code> layer.
Within the counts layer, you will find the sum of the gene expression
for each gene across all cells from an individual (grouped on
<code>sample_identifier</code>). The <code>pct.expression</code> layer
is the percentage of cells from that individual that express the gene.
For instance, if 1 out of 10 cells from a subject express a gene, the
<code>pct.expression</code> will be 0.1.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#inspecting counts. The matrices are sparse, so zeroes are represented with the '.' character.</span></span>
<span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">pbulk</span>, layer <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; 3 x 3 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;        control|baseline|Subject.1.control.baseline</span></span>
<span><span class="co">#&gt; SAMD11                                           .</span></span>
<span><span class="co">#&gt; NOC2L                                            .</span></span>
<span><span class="co">#&gt; KLHL17                                           .</span></span>
<span><span class="co">#&gt;        control|baseline|Subject.2.control.baseline</span></span>
<span><span class="co">#&gt; SAMD11                                           .</span></span>
<span><span class="co">#&gt; NOC2L                                            1</span></span>
<span><span class="co">#&gt; KLHL17                                           .</span></span>
<span><span class="co">#&gt;        control|baseline|Subject.3.control.baseline</span></span>
<span><span class="co">#&gt; SAMD11                                           .</span></span>
<span><span class="co">#&gt; NOC2L                                            1</span></span>
<span><span class="co">#&gt; KLHL17                                           1</span></span>
<span></span>
<span><span class="co">#inspecting pct.expression. This is also sparse, so unexpressed genes in a sample are represented with the '.' character.</span></span>
<span><span class="fu">GetAssayData</span><span class="op">(</span><span class="va">pbulk</span>, layer <span class="op">=</span> <span class="st">"pct.expression"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt; 3 x 3 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">#&gt;        control|baseline|Subject.1.control.baseline</span></span>
<span><span class="co">#&gt; SAMD11                                           .</span></span>
<span><span class="co">#&gt; NOC2L                                            .</span></span>
<span><span class="co">#&gt; KLHL17                                           .</span></span>
<span><span class="co">#&gt;        control|baseline|Subject.2.control.baseline</span></span>
<span><span class="co">#&gt; SAMD11                                  .         </span></span>
<span><span class="co">#&gt; NOC2L                                   0.02941176</span></span>
<span><span class="co">#&gt; KLHL17                                  .         </span></span>
<span><span class="co">#&gt;        control|baseline|Subject.3.control.baseline</span></span>
<span><span class="co">#&gt; SAMD11                                  .         </span></span>
<span><span class="co">#&gt; NOC2L                                   0.02941176</span></span>
<span><span class="co">#&gt; KLHL17                                  0.02941176</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="differential-expression-what-is-a-contrast">Differential Expression: What is a ‘Contrast’?<a class="anchor" aria-label="anchor" href="#differential-expression-what-is-a-contrast"></a>
</h2>
<p>Contrasts are comparisons of samples. When one discusses differential
expression between two groups, that comparison is performed via a
“contrast” between those two groups. To link which groups will be
compared to one another, we use a design matrix, which allows us to
impose our experimental design onto the samples and group them
appropriately to determine gene expression differences between the
groups.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignModelMatrix.html">DesignModelMatrix</a></span><span class="op">(</span><span class="va">pbulk</span>, contrast_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vaccine_cohort"</span><span class="op">)</span>, sampleIdCol <span class="op">=</span> <span class="st">"sample_identifier"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">design</span>, </span>
<span>                        cluster_columns <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                        cluster_rows <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                        show_row_names <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                        column_title <span class="op">=</span> <span class="st">"Experimental Groups"</span>,</span>
<span>                        row_title <span class="op">=</span> <span class="st">"Individual Sample Labels"</span>,</span>
<span>                        col <span class="op">=</span> <span class="fu">colorRamp2</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cadetblue2"</span>, <span class="st">"coral"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        rect_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>col <span class="op">=</span> <span class="st">"black"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>                        heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Group Membership"</span>, </span>
<span>                                                    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, </span>
<span>                                                    color_bar <span class="op">=</span> <span class="st">"discrete"</span>, </span>
<span>                                                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Not in Group"</span>, <span class="st">"In Group"</span><span class="op">)</span></span>
<span>                                                    <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/example%20design%20matrix%201-1.png" width="960"></p>
<p>However, you can make more complicated designs that incorporate
multiple levels of experimental variables. Fitting a model downstream
will require at minimum more than 1 data point per group, but more
biological replication is better.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignModelMatrix.html">DesignModelMatrix</a></span><span class="op">(</span><span class="va">pbulk</span>, contrast_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vaccine_cohort"</span>, <span class="st">"timepoint"</span><span class="op">)</span>, sampleIdCol <span class="op">=</span> <span class="st">"sample_identifier"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">design</span>, </span>
<span>                        cluster_columns <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                        cluster_rows <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                        show_row_names <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                        column_title <span class="op">=</span> <span class="st">"Experimental Groups"</span>,</span>
<span>                        row_title <span class="op">=</span> <span class="st">"Individual Sample Labels"</span>,</span>
<span>                        col <span class="op">=</span> <span class="fu">colorRamp2</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cadetblue2"</span>, <span class="st">"coral"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        rect_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>col <span class="op">=</span> <span class="st">"black"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>                        heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Group Membership"</span>, </span>
<span>                                                    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, </span>
<span>                                                    color_bar <span class="op">=</span> <span class="st">"discrete"</span>, </span>
<span>                                                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Not in Group"</span>, <span class="st">"In Group"</span><span class="op">)</span></span>
<span>                                                    <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/example%20design%20matrix%202-1.png" width="960"></p>
</div>
<div class="section level2">
<h2 id="running-a-contrast-vaccineone-vs-vaccinetwo">Running a Contrast: vaccineOne vs vaccineTwo<a class="anchor" aria-label="anchor" href="#running-a-contrast-vaccineone-vs-vaccinetwo"></a>
</h2>
<p>Now that you have the sample grouping established, we can fit models
to our data and compare those to calculate log fold changes and
statistical significance. Normally, these models are regressions using
some kind of count-based distribution (poisson, quasipoisson, negative
binomial) using packages such as DESeq2 or edgeR.</p>
<p>Let’s define a contrast using just vaccine information and compare
vaccine one and vaccine two:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignModelMatrix.html">DesignModelMatrix</a></span><span class="op">(</span><span class="va">pbulk</span>, contrast_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vaccine_cohort"</span><span class="op">)</span>, sampleIdCol <span class="op">=</span> <span class="st">"sample_identifier"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ComplexHeatmap/man/Heatmap.html" class="external-link">Heatmap</a></span><span class="op">(</span><span class="va">design</span>, </span>
<span>                        cluster_columns <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                        cluster_rows <span class="op">=</span> <span class="cn">F</span>, </span>
<span>                        show_row_names <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                        column_title <span class="op">=</span> <span class="st">"Experimental Groups"</span>,</span>
<span>                        row_title <span class="op">=</span> <span class="st">"Individual Sample Labels"</span>,</span>
<span>                        col <span class="op">=</span> <span class="fu">colorRamp2</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"cadetblue2"</span>, <span class="st">"coral"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                        rect_gp <span class="op">=</span> <span class="fu">gpar</span><span class="op">(</span>col <span class="op">=</span> <span class="st">"black"</span>, lwd <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>,</span>
<span>                        heatmap_legend_param <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Group Membership"</span>, </span>
<span>                                                    at <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span>, </span>
<span>                                                    color_bar <span class="op">=</span> <span class="st">"discrete"</span>, </span>
<span>                                                    labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Not in Group"</span>, <span class="st">"In Group"</span><span class="op">)</span></span>
<span>                                                    <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/design%20matrix%20for%20contrasts-1.png" width="960"></p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#fit a GLM using edgeR</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PerformGlmFit.html">PerformGlmFit</a></span><span class="op">(</span><span class="va">pbulk</span>, design <span class="op">=</span> <span class="va">design</span>, test.use <span class="op">=</span> <span class="st">"QLF"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/example%20contrast-1.png" width="700"></p>
<pre><code><span><span class="co">#&gt; NULL</span></span>
<span></span>
<span><span class="co"># Define Contrast using the notation Group1 - Group2. </span></span>
<span><span class="co"># Genes that are upregulated in group 1 will have a positive log fold change</span></span>
<span><span class="co"># while genes upregulated in group 2 will have a negative log fold change.</span></span>
<span><span class="va">contrast_name</span> <span class="op">&lt;-</span> <span class="st">"vaccineOne - vaccineTwo"</span></span>
<span><span class="va">contrast</span> <span class="op">&lt;-</span></span>
<span>  <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/makeContrasts.html" class="external-link">makeContrasts</a></span><span class="op">(</span>contrasts <span class="op">=</span> <span class="va">contrast_name</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">result</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PerformDifferentialExpression.html">PerformDifferentialExpression</a></span><span class="op">(</span></span>
<span>  <span class="va">fit</span>,</span>
<span>  <span class="va">contrast</span>,</span>
<span>  <span class="va">contrast_name</span>,</span>
<span>  logFC_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  FDR_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  test.use <span class="op">=</span> <span class="st">"QLF"</span>,</span>
<span>  showPlots <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/example%20contrast-2.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/example%20contrast-3.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="great-now-let-us-automate-it-">Great, now let us automate it.<a class="anchor" aria-label="anchor" href="#great-now-let-us-automate-it-"></a>
</h2>
<p>In larger studies where you have several experimental variables, the
number of possible comparisons grow combinatorially.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#single level design matrix</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignModelMatrix.html">DesignModelMatrix</a></span><span class="op">(</span><span class="va">pbulk</span>, contrast_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vaccine_cohort"</span><span class="op">)</span>, sampleIdCol <span class="op">=</span> <span class="st">"sample_identifier"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#6 possible contrasts with a single level design</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span>,m <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 6</span></span>
<span></span>
<span><span class="co">#two-level design matrix</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/DesignModelMatrix.html">DesignModelMatrix</a></span><span class="op">(</span><span class="va">pbulk</span>, contrast_columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"vaccine_cohort"</span>, <span class="st">"timepoint"</span><span class="op">)</span>, sampleIdCol <span class="op">=</span> <span class="st">"sample_identifier"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#66 possible contrasts by including timepoint in our contrasts...</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/combn.html" class="external-link">combn</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">design</span><span class="op">)</span>,m <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 66</span></span></code></pre></div>
<p>However, you probably don’t care about all of the possible
comparisons. If your design matrix includes cell type, timepoint, and
vaccine, there will be many differentially expressed genes between B
cells and Myeloid cells, but not due to timepoint or vaccine cohort.
Let’s use some logic to specify exactly what we want to measure across
all possible pseudobulking contrasts.</p>
<p>Here, I’ll use the two-level study design, but I’ve learned that
vaccineTwo recently failed its clinical trial, so I only want to
investigate vaccineOne over my time course. However, I also know that
the effect of the vaccine could vary over time, so I want to make sure
that I only compare the vaccineOne to either unvax or control within a
given timepoint.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">logicList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"vaccine_cohort"</span>, <span class="st">"xor"</span>, <span class="st">"vaccineOne"</span><span class="op">)</span>, </span>
<span>                   <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"vaccine_cohort"</span>, <span class="st">"nor"</span>, <span class="st">"vaccineTwo"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#We can easily visualize the list of lists as a dataframe to inspect the logic gates.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="st">"rbind"</span>, <span class="va">logicList</span><span class="op">)</span></span>
<span><span class="co">#&gt;      [,1]             [,2]  [,3]        </span></span>
<span><span class="co">#&gt; [1,] "vaccine_cohort" "xor" "vaccineOne"</span></span>
<span><span class="co">#&gt; [2,] "vaccine_cohort" "nor" "vaccineTwo"</span></span></code></pre></div>
<p>This list of lists sets up our logic gates for filtering contrasts.
One of the sides of the contrast needs to be vaccineOne (XOR gate on
vaccineOne), and neither of the contrasts can have vaccineTwo in them
(NOR gate on vaccineTwo).</p>
<p>To accomplish ensuring that the timepoint variable remains fixed
across the contrasts, we can pass <code>'timepoint'</code> to the
<code>'requireIdenticalFields'</code> argument in the
<code>FilterPseudobulkContrasts</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">filtered_contrasts</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/FilterPseudobulkContrasts.html">FilterPseudobulkContrasts</a></span><span class="op">(</span>logicList <span class="op">=</span> <span class="va">logicList</span>, </span>
<span>                                                          design <span class="op">=</span> <span class="va">design</span>, </span>
<span>                                                          useRequireIdenticalLogic <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                                                          requireIdenticalFields <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"timepoint"</span><span class="op">)</span>, </span>
<span>                                                          filteredContrastsOutputFile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">filtered_contrasts</span></span>
<span><span class="co">#&gt;   positive_contrast_vaccine_cohort positive_contrast_timepoint</span></span>
<span><span class="co">#&gt; 1                          control                    baseline</span></span>
<span><span class="co">#&gt; 2                          control                        day4</span></span>
<span><span class="co">#&gt; 3                          control                    necropsy</span></span>
<span><span class="co">#&gt; 4                            unvax                    baseline</span></span>
<span><span class="co">#&gt; 5                            unvax                        day4</span></span>
<span><span class="co">#&gt; 6                            unvax                    necropsy</span></span>
<span><span class="co">#&gt;   negative_contrast_vaccine_cohort negative_contrast_timepoint</span></span>
<span><span class="co">#&gt; 1                       vaccineOne                    baseline</span></span>
<span><span class="co">#&gt; 2                       vaccineOne                        day4</span></span>
<span><span class="co">#&gt; 3                       vaccineOne                    necropsy</span></span>
<span><span class="co">#&gt; 4                       vaccineOne                    baseline</span></span>
<span><span class="co">#&gt; 5                       vaccineOne                        day4</span></span>
<span><span class="co">#&gt; 6                       vaccineOne                    necropsy</span></span></code></pre></div>
<p>Now we can see that we have six contrasts (down from 66) that compare
vaccineOne to either control or unvax, within a fixed timepoint.</p>
</div>
<div class="section level2">
<h2 id="run-filtered-contrasts">Run Filtered Contrasts<a class="anchor" aria-label="anchor" href="#run-filtered-contrasts"></a>
</h2>
<p>We can use the <code>RunFilteredContrasts</code> function to execute
the basic edgeR QLF differential expression function on each of these
contrasts. This will print the biological coefficient of variation plot
for each subset so you can grossly interpret whether or not the model
fitting is suitable.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DE_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunFilteredContrasts.html">RunFilteredContrasts</a></span><span class="op">(</span>seuratObj <span class="op">=</span> <span class="va">pbulk</span>, </span>
<span>                       filteredContrastsDataframe <span class="op">=</span> <span class="va">filtered_contrasts</span>, </span>
<span>                       design <span class="op">=</span> <span class="va">design</span>,</span>
<span>                       test.use <span class="op">=</span> <span class="st">"QLF"</span>, </span>
<span>                       logFC_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       FDR_threshold <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                       minCountsPerGene <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                       assayName <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/run%20filtered%20contrasts-1.png" width="288"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/run%20filtered%20contrasts-2.png" width="288"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/run%20filtered%20contrasts-3.png" width="288"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/run%20filtered%20contrasts-4.png" width="288"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/run%20filtered%20contrasts-5.png" width="288"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/run%20filtered%20contrasts-6.png" width="288"></p>
<pre><code><span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_day4-vaccineOne_day4"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_day4-vaccineOne_day4"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_necropsy-vaccineOne_necropsy"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_necropsy-vaccineOne_necropsy"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  unvax_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax_day4-vaccineOne_day4"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  unvax_day4-vaccineOne_day4"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax_necropsy-vaccineOne_necropsy"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  unvax_necropsy-vaccineOne_necropsy"</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre>
<div class="section level3">
<h3 id="thresholds-are-for-optional-plotting-in-runfilteredcontrasts-but-are-used-for-filtering-in-pseudobulkingbarplot">Thresholds are for optional plotting in
<code>RunFilteredContrasts</code>, but are used for filtering in
<code>PseudobulkingBarPlot</code><a class="anchor" aria-label="anchor" href="#thresholds-are-for-optional-plotting-in-runfilteredcontrasts-but-are-used-for-filtering-in-pseudobulkingbarplot"></a>
</h3>
<p>Please note that the logFC and FDR thresholds here are for displaying
the volcano plots (if desired) only, and they perform no filtering on
genes in the <code>DE_results</code> list.</p>
<p>Permissive thresholds (on one contrast):</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">permissive_contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunFilteredContrasts.html">RunFilteredContrasts</a></span><span class="op">(</span>seuratObj <span class="op">=</span> <span class="va">pbulk</span>, </span>
<span>                       filteredContrastsDataframe <span class="op">=</span> <span class="va">filtered_contrasts</span><span class="op">[</span><span class="fl">4</span>,<span class="op">]</span>, </span>
<span>                       design <span class="op">=</span> <span class="va">design</span>,</span>
<span>                       test.use <span class="op">=</span> <span class="st">"QLF"</span>, </span>
<span>                       logFC_threshold <span class="op">=</span> <span class="fl">0</span>,</span>
<span>                       FDR_threshold <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                       minCountsPerGene <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                       assayName <span class="op">=</span> <span class="st">"RNA"</span>, </span>
<span>                       showPlots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Permissive%20Thresholds-1.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Permissive%20Thresholds-2.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Permissive%20Thresholds-3.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  unvax_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre>
<p>Stricter thresholds (on one contrast):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">strict_contrast</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/RunFilteredContrasts.html">RunFilteredContrasts</a></span><span class="op">(</span>seuratObj <span class="op">=</span> <span class="va">pbulk</span>, </span>
<span>                       filteredContrastsDataframe <span class="op">=</span> <span class="va">filtered_contrasts</span><span class="op">[</span><span class="fl">4</span>,<span class="op">]</span>, </span>
<span>                       design <span class="op">=</span> <span class="va">design</span>,</span>
<span>                       test.use <span class="op">=</span> <span class="st">"QLF"</span>, </span>
<span>                       logFC_threshold <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       FDR_threshold <span class="op">=</span> <span class="fl">1e-30</span>,</span>
<span>                       minCountsPerGene <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                       assayName <span class="op">=</span> <span class="st">"RNA"</span>, </span>
<span>                       showPlots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Strict%20Thresholds-1.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Strict%20Thresholds-2.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Strict%20Thresholds-3.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  unvax_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  unvax_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; `stat_bin()` using `bins = 30`. Pick better value with `binwidth`.</span></span></code></pre>
<p>The results of these fits are collected in an indexed list.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">DE_results</span><span class="op">$</span><span class="va">`1`</span><span class="op">)</span></span>
<span><span class="co">#&gt;                         logFC    logCPM        F       PValue          FDR</span></span>
<span><span class="co">#&gt; ENSMMUG00000013779  2.5050954  7.947134 41.42212 2.343566e-09 0.0000210546</span></span>
<span><span class="co">#&gt; PPBP                5.9204304  6.281533 26.74732 9.143648e-07 0.0041073267</span></span>
<span><span class="co">#&gt; ENSMMUG00000045466  4.9622553  5.720617 18.99011 2.745347e-05 0.0711728374</span></span>
<span><span class="co">#&gt; MYL9                5.0419674  5.761552 18.67463 3.168871e-05 0.0711728374</span></span>
<span><span class="co">#&gt; CCL4                1.1331389  8.081421 17.09495 6.449480e-05 0.1158842625</span></span>
<span><span class="co">#&gt; FTH1               -0.5151792 11.954005 16.30541 9.309469e-05 0.1393937809</span></span>
<span><span class="co">#&gt;                                  gene       pct.1      pct.2</span></span>
<span><span class="co">#&gt; ENSMMUG00000013779 ENSMMUG00000013779 0.100000000 0.07751938</span></span>
<span><span class="co">#&gt; PPBP                             PPBP 0.007692308 0.00000000</span></span>
<span><span class="co">#&gt; ENSMMUG00000045466 ENSMMUG00000045466 0.053846154 0.00000000</span></span>
<span><span class="co">#&gt; MYL9                             MYL9 0.030769231 0.00000000</span></span>
<span><span class="co">#&gt; CCL4                             CCL4 0.246153846 0.16279070</span></span>
<span><span class="co">#&gt; FTH1                             FTH1 0.892307692 0.95348837</span></span>
<span><span class="co">#&gt;                                           contrast_name positive_contrast</span></span>
<span><span class="co">#&gt; ENSMMUG00000013779 control_baseline-vaccineOne_baseline  control_baseline</span></span>
<span><span class="co">#&gt; PPBP               control_baseline-vaccineOne_baseline  control_baseline</span></span>
<span><span class="co">#&gt; ENSMMUG00000045466 control_baseline-vaccineOne_baseline  control_baseline</span></span>
<span><span class="co">#&gt; MYL9               control_baseline-vaccineOne_baseline  control_baseline</span></span>
<span><span class="co">#&gt; CCL4               control_baseline-vaccineOne_baseline  control_baseline</span></span>
<span><span class="co">#&gt; FTH1               control_baseline-vaccineOne_baseline  control_baseline</span></span>
<span><span class="co">#&gt;                      negative_contrast positive_contrast_vaccine_cohort</span></span>
<span><span class="co">#&gt; ENSMMUG00000013779 vaccineOne_baseline                          control</span></span>
<span><span class="co">#&gt; PPBP               vaccineOne_baseline                          control</span></span>
<span><span class="co">#&gt; ENSMMUG00000045466 vaccineOne_baseline                          control</span></span>
<span><span class="co">#&gt; MYL9               vaccineOne_baseline                          control</span></span>
<span><span class="co">#&gt; CCL4               vaccineOne_baseline                          control</span></span>
<span><span class="co">#&gt; FTH1               vaccineOne_baseline                          control</span></span>
<span><span class="co">#&gt;                    positive_contrast_timepoint negative_contrast_vaccine_cohort</span></span>
<span><span class="co">#&gt; ENSMMUG00000013779                    baseline                       vaccineOne</span></span>
<span><span class="co">#&gt; PPBP                                  baseline                       vaccineOne</span></span>
<span><span class="co">#&gt; ENSMMUG00000045466                    baseline                       vaccineOne</span></span>
<span><span class="co">#&gt; MYL9                                  baseline                       vaccineOne</span></span>
<span><span class="co">#&gt; CCL4                                  baseline                       vaccineOne</span></span>
<span><span class="co">#&gt; FTH1                                  baseline                       vaccineOne</span></span>
<span><span class="co">#&gt;                    negative_contrast_timepoint</span></span>
<span><span class="co">#&gt; ENSMMUG00000013779                    baseline</span></span>
<span><span class="co">#&gt; PPBP                                  baseline</span></span>
<span><span class="co">#&gt; ENSMMUG00000045466                    baseline</span></span>
<span><span class="co">#&gt; MYL9                                  baseline</span></span>
<span><span class="co">#&gt; CCL4                                  baseline</span></span>
<span><span class="co">#&gt; FTH1                                  baseline</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="pseudobulking-bar-plot">Pseudobulking Bar Plot<a class="anchor" aria-label="anchor" href="#pseudobulking-bar-plot"></a>
</h2>
<p>We can visualize the results of <code>RunFilteredContrasts</code> in
a bar plot, showing the magnitude of up- and down-regulated genes over
our contrasts that address our hypothesis. The colors indicate whether
or not the genes are unique (red or dark blue) to their contrast or
shared across the contrasts (orange, light blue). Please note, the
thresholds are used to determine if the genes are differentially
expressed or not within each contrast.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barplot_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PseudobulkingBarPlot.html">PseudobulkingBarPlot</a></span><span class="op">(</span><span class="va">DE_results</span>, </span>
<span>                                        title <span class="op">=</span> <span class="st">"Vaccine One vs Unvaxxed/Control"</span>, </span>
<span>                                        logFC_threshold <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                        FDR_threshold <span class="op">=</span> <span class="fl">.5</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Create%20Permissive%20Bar%20Plot-1.png" width="700"></p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">barplot_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PseudobulkingBarPlot.html">PseudobulkingBarPlot</a></span><span class="op">(</span><span class="va">DE_results</span>, </span>
<span>                                        title <span class="op">=</span> <span class="st">"Vaccine One vs Unvaxxed/Control"</span>, </span>
<span>                                        logFC_threshold <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>                                        FDR_threshold <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Create%20Strict%20Bar%20Plot-1.png" width="700"></p>
<p>Further annotation of the bar plot requires some custom code, but can
be formatted like so using the data frame returned from
<code>PseudobulkingBarPlot</code>:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">#inspect dataframe column names to determine how the fill aesthetic needs to be specified for the geom_tile() calls. </span></span>
<span><span class="va">barplot_results</span><span class="op">$</span><span class="va">dataframe</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "contrast_name"                     "positive_contrast"                </span></span>
<span><span class="co">#&gt;  [3] "positive_contrast_vaccine_cohort"  "positive_contrast_timepoint"      </span></span>
<span><span class="co">#&gt;  [5] "negative_contrast"                 "negative_contrast_vaccine_cohort" </span></span>
<span><span class="co">#&gt;  [7] "negative_contrast_timepoint"       "number_of_positive_nonunique_DEGs"</span></span>
<span><span class="co">#&gt;  [9] "number_of_positive_unique_DEGs"    "number_of_negative_nonunique_DEGs"</span></span>
<span><span class="co">#&gt; [11] "number_of_negative_unique_DEGs"    "total_DEGs"                       </span></span>
<span><span class="co">#&gt; [13] "DEG_Magnitude"</span></span>
<span></span>
<span><span class="va">metadata_tiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">barplot_results</span><span class="op">$</span><span class="va">dataframe</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="co">#define a custom scale for timepoint metadata</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html" class="external-link">n_distinct</a></span><span class="op">(</span><span class="va">pbulk</span><span class="op">$</span><span class="va">timepoint</span><span class="op">)</span>, palette <span class="op">=</span> <span class="st">"Purples 3"</span><span class="op">)</span><span class="op">)</span>, guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>order <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Timepoint"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">contrast_name</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">total_DEGs</span><span class="op">)</span><span class="op">)</span>, y <span class="op">=</span> <span class="fl">1</span>, fill <span class="op">=</span> <span class="va">barplot_results</span><span class="op">$</span><span class="va">dataframe</span><span class="op">$</span><span class="va">negative_contrast_timepoint</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">contrast_name</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">total_DEGs</span><span class="op">)</span><span class="op">)</span>, y <span class="op">=</span> <span class="fl">2</span>, fill <span class="op">=</span> <span class="va">barplot_results</span><span class="op">$</span><span class="va">dataframe</span><span class="op">$</span><span class="va">positive_contrast_timepoint</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggnewscale</span><span class="fu">::</span><span class="fu"><a href="https://eliocamp.github.io/ggnewscale/reference/new_scale.html" class="external-link">new_scale_fill</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>  <span class="co">#define a custom scale for vaccine_cohort metadata</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu">grDevices</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/n_distinct.html" class="external-link">n_distinct</a></span><span class="op">(</span><span class="va">pbulk</span><span class="op">$</span><span class="va">vaccine_cohort</span><span class="op">)</span>, palette <span class="op">=</span> <span class="st">"Inferno"</span><span class="op">)</span><span class="op">)</span>, guide <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>order <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"Vaccine"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">contrast_name</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">total_DEGs</span><span class="op">)</span><span class="op">)</span>, y <span class="op">=</span> <span class="fl">3</span>, fill <span class="op">=</span> <span class="va">barplot_results</span><span class="op">$</span><span class="va">dataframe</span><span class="op">$</span><span class="va">negative_contrast_vaccine_cohort</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_tile.html" class="external-link">geom_tile</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">contrast_name</span>, <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">total_DEGs</span><span class="op">)</span><span class="op">)</span>, y <span class="op">=</span> <span class="fl">4</span>, fill <span class="op">=</span> <span class="va">barplot_results</span><span class="op">$</span><span class="va">dataframe</span><span class="op">$</span><span class="va">positive_contrast_vaccine_cohort</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggnewscale</span><span class="fu">::</span><span class="fu"><a href="https://eliocamp.github.io/ggnewscale/reference/new_scale.html" class="external-link">new_scale_fill</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  </span>
<span>  <span class="co">#match the formatting of the Pseudobulking Bar Plot</span></span>
<span>  <span class="fu">egg</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/egg/man/theme_article.html" class="external-link">theme_article</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span>  <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Contrast"</span><span class="op">)</span><span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Metadata Variables"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="co">#add some dashed lines to delineate metadata groupings</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">4.5</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"black"</span>, linetype <span class="op">=</span> <span class="st">"dashed"</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">barplot_results</span><span class="op">$</span><span class="va">dataframe</span><span class="op">$</span><span class="va">contrast_name</span><span class="op">)</span><span class="op">)</span><span class="op">+</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_discrete.html" class="external-link">scale_y_discrete</a></span><span class="op">(</span>limits <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rev.html" class="external-link">rev</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Positive Contrast Vaccine"</span>, <span class="st">"Negative Contrast Vaccine"</span>,<span class="st">"Positive Contrast Timepoint"</span>,<span class="st">"Negative Contrast Timepoint"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">DEG_Magnitude</span>, scales <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">metadata_tiles</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Annotate%20bar%20Plot-1.png" width="700"></p>
<p>Then, you can add these plots together using patchwork to annotate
the contrasts shown in the Pseudobulking Bar Plot.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span></span>
<span><span class="op">(</span><span class="va">barplot_results</span><span class="op">$</span><span class="va">barPlot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op">/</span> <span class="op">(</span><span class="va">metadata_tiles</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Differential Expression Contrasts"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">DEG_Magnitude</span>, scales <span class="op">=</span> <span class="st">"free_x"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">'collect'</span><span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Show%20Final%20Bar%20Plot-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="differential-expression-heatmap">Differential Expression Heatmap<a class="anchor" aria-label="anchor" href="#differential-expression-heatmap"></a>
</h2>
<p>Very often, we would like to interpret a group of genes’ (module)
expression relative to our experimental design. We can visualize gene
groupings using a carefully formatted heatmap. For this heatmap, we’d
like to show genes that are upregulated in our experimental condition
compared to our control.</p>
<p>First, let’s reuse the results of
<code>FilterPseudobulkingContrasts</code> and apply some differential
gene expression cutoffs so we’re working with a smaller gene space. You
can also use other feature selection algorithms such as CellMembrane’s
<code><a href="../reference/FitRegularizedClassificationGlm.html">FitRegularizedClassificationGlm()</a></code> or other machine learning
approaches to determine appropriate gene sets.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#bind the list into a dataframe using thresholds to determine the differentially expressed genes</span></span>
<span><span class="va">DEGs</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">DE_results</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#select all of the genes that are differentially expressed in any contrast across the transcriptome</span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">DEGs</span><span class="op">$</span><span class="va">gene</span><span class="op">)</span></span></code></pre></div>
<p>One can also filter based on the number of cells that express the
gene in each contrast. You can use this to bias towards genes that are
expressed in a larger proportion of cells in one of the experimental
conditions (i.e. “marker genes”), or to ensure that the gene is
expressed in many cells, rather than acutely in a few cells. Within the
<code>DE_results</code> dataframes, <code>pct.1</code> refers to the
proportion of cells that express the gene in the first group of the
contrast (i.e. positive contrast), and <code>pct.2</code> refers to the
proportion of cells that express the gene in the second group of the
contrast (i.e. negative contrast).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">DEGs_no_rare_cells</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">DE_results</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pct.1</span> <span class="op">&gt;</span> <span class="fl">0.1</span> <span class="op">&amp;</span> <span class="va">pct.2</span> <span class="op">&gt;</span> <span class="fl">0.1</span><span class="op">)</span> <span class="co"># an example of filtering for genes that have at least 10% of cells expressing the gene in each group.</span></span>
<span></span>
<span><span class="va">DEGs_more_common_in_positive_contrast</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_rows.html" class="external-link">bind_rows</a></span><span class="op">(</span><span class="va">DE_results</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">logFC</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">FDR</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">pct.1</span> <span class="op">-</span> <span class="va">pct.2</span> <span class="op">&gt;</span> <span class="fl">0.25</span><span class="op">)</span> <span class="co"># an example of filtering for genes that have at least 25% more cells expressing the gene in the positive contrast than the negative contrast.</span></span></code></pre></div>
<div class="section level3">
<h3 id="create-the-heatmap-on-just-the-degs">Create the heatmap on just the DEGs<a class="anchor" aria-label="anchor" href="#create-the-heatmap-on-just-the-degs"></a>
</h3>
<p>Then, we can make a heatmap on just the differentially expressed
genes so we can interpret their expression in relation to our vaccine
groups.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heatmap_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PseudobulkingDEHeatmap.html">PseudobulkingDEHeatmap</a></span><span class="op">(</span>seuratObj <span class="op">=</span> <span class="va">pbulk</span>, </span>
<span>                       geneSpace <span class="op">=</span> <span class="va">genes</span>, </span>
<span>                       contrastField <span class="op">=</span> <span class="st">"vaccine_cohort"</span>, </span>
<span>                       negativeContrastValue <span class="op">=</span> <span class="st">"control"</span>, </span>
<span>                       sampleIdCol <span class="op">=</span> <span class="st">'sample_identifier'</span></span>
<span>                       <span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Single%20Variable%20Heatmap-1.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Single%20Variable%20Heatmap-2.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Single%20Variable%20Heatmap-3.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-unvax"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control-unvax"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control-vaccineOne"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineTwo"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control-vaccineTwo"</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre>
<p>We can compare the gene log fold changes over the vaccine cohorts and
find gene modules that are consistently upregulated.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heatmap_list</span><span class="op">$</span><span class="va">heatmap</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/show%20first%20heatmap-1.png" width="700"></p>
</div>
<div class="section level3">
<h3 id="de-heatmap-using-two-variables">DE Heatmap Using Two Variables<a class="anchor" aria-label="anchor" href="#de-heatmap-using-two-variables"></a>
</h3>
<p>Often, you’ll have a treatment and time course, where you measure
transcriptomics in subjects over time after treatment. We can pass that
information as a <code>subgroupingVariable</code> to
<code>PseudobulkingDEHeatmap</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heatmap_list</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/PseudobulkingDEHeatmap.html">PseudobulkingDEHeatmap</a></span><span class="op">(</span>seuratObj <span class="op">=</span> <span class="va">pbulk</span>, </span>
<span>                       geneSpace <span class="op">=</span> <span class="va">genes</span>, </span>
<span>                       contrastField <span class="op">=</span> <span class="st">"vaccine_cohort"</span>, </span>
<span>                       negativeContrastValue <span class="op">=</span> <span class="st">"control"</span>, </span>
<span>                       subgroupingVariable <span class="op">=</span> <span class="st">'timepoint'</span>,</span>
<span>                       sampleIdCol <span class="op">=</span> <span class="st">'sample_identifier'</span></span>
<span>                       <span class="op">)</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-1.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-2.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-3.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-4.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-5.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-6.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-7.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-8.png" width="700"><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/Two%20Variable%20Heatmap-9.png" width="700"></p>
<pre><code><span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-unvax"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_baseline-unvax_baseline"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_baseline-unvax_baseline"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_baseline-vaccineOne_baseline"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineTwo"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_baseline-vaccineTwo_baseline"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_baseline-vaccineTwo_baseline"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-unvax"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_day4-unvax_day4"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_day4-unvax_day4"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_day4-vaccineOne_day4"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_day4-vaccineOne_day4"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineTwo"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_day4-vaccineTwo_day4"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_day4-vaccineTwo_day4"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-unvax"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_necropsy-unvax_necropsy"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_necropsy-unvax_necropsy"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineOne"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_necropsy-vaccineOne_necropsy"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_necropsy-vaccineOne_necropsy"</span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; [1] "Contrast columns: vaccine_cohort" "Contrast columns: timepoint"     </span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control-vaccineTwo"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "24"              </span></span>
<span><span class="co">#&gt; [1] "colsums: 12" "colsums: 12"</span></span>
<span><span class="co">#&gt; [1] "Subsetting Cells for:  control_necropsy-vaccineTwo_necropsy"</span></span>
<span><span class="co">#&gt; [1] "Filtering cells..."</span></span>
<span><span class="co">#&gt; [1] "Inspecting assay layers: RNA"</span></span>
<span><span class="co">#&gt; [1] "Layers: counts.1,counts.2,data.1,scale.data.1,pct.expression.1,data.2,scale.data.2,pct.expression.2"</span></span>
<span><span class="co">#&gt; [1] "Class: Assay5"</span></span>
<span><span class="co">#&gt; [1] "Joining layers: RNA, counts;data;scale.data;pct.expression"</span></span>
<span><span class="co">#&gt; [1] "After join: pct.expression,scale.data,data,counts"</span></span>
<span><span class="co">#&gt; [1] "Remaining Cells:" "8"               </span></span>
<span><span class="co">#&gt; [1] "colsums: 8"</span></span>
<span><span class="co">#&gt; [1] "Performing DE for:  control_necropsy-vaccineTwo_necropsy"</span></span>
<span><span class="co">#&gt; NULL</span></span></code></pre>
<p>Now, we can interpret the changes within vaccine cohorts and over
time!</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">heatmap_list</span><span class="op">$</span><span class="va">heatmap</span></span></code></pre></div>
<p><img src="Pseudobulking-and-Filtered-Contrasts_files/figure-html/show%20second%20heatmap-1.png" width="768"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by The package maintainer.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer>
</div>






  </body>
</html>
